torch/testing/_internal/common_utils.py

```diff
@@ -777,27 +777,65 @@ def safeCoalesce(self, t):
 
         return tg
 
-    def assertEqual(self, x, y, prec=None, message='', allow_inf=False, exact_dtype=None):
+    # Some analysis of tolerance by logging tests from test_torch.py can be found
+    # in https://github.com/pytorch/pytorch/pull/32538.
+    dtype_precisions = {
+        torch.half: 0.001,
+        torch.bfloat16: 0.016,
+        torch.float: 1.3e-6,
+        torch.double: 5e-7,
+        torch.int: 0
+    }
+
+    def compared_dtype(self, x, y):
+        dtype = torch.float
+        if isinstance(x, torch.Tensor):
+            dtype = x.dtype
+        elif isinstance(y, torch.Tensor):
+            dtype = y.dtype
+        elif TEST_NUMPY and isinstance(x, numpy.ndarray):
+            dtype = torch.from_numpy(x).dtype
+
+        if dtype.is_floating_point:
+            return dtype
+        else:
+            return torch.int
+
+
+    def assertEqual(self, x, y, message='', rtol=None, atol=None, allow_inf=False, exact_dtype=None):
+        # we allow setting an absolute tolerance as a positional arg for BC with legacy testing behavior.
+        if isinstance(message, Number):
+            self.assertIsNone(atol, "don't combine positional prec and atol")
```

Comment:
If the old user wrote `assertEqual(x, y, tol, "foo bar")` they'll get a weird error message here, right?

Example:
```
test/test_torch.py:2471:        self.assertEqual(target_value, forked_value, 0, "RNG has not forked correctly.")
```

Reply:
That's correct. test_torch.py:2471 was changed to:
        `self.assertEqual(target_value, forked_value, atol=0, message="RNG has not forked correctly.")`
in this PR

Actually though, I'll add a unit test to ensure that calls in that form do in fact fail.